### ISSUE TITLE
Proposal for new Security Group rules

### ISSUE THREAD

**Issue by kddejong** [aws-cloudformation/cfn-python-lint#1413]
Proposing some new rules for Security Groups.

- Don't  use both resource `AWS::EC2::SecurityGroup` property `SecurityGroupEgress` along with `AWS::EC2::SecurityGroupEgress`.  Equivalent for Ingress.  This is to help prevent confusion between the two methodologies to specify Security Group rules.  Error level is **Warning**.  I think we should favor the `AWS::EC2::SecurityGroupEgress` methodology.  
- When specifying an `Egress` rule on a security group the default egress rule is automatically removed.  If you remove that security group rule the default egress rule isn't automatically added back.  Error level is **Info**.  This rule shouldn't change the exit code.

**Comment by Tro95** [comment]
I like this idea, but particularly for the first point I'm not sure whether **Warning** is the right way to surface this. I feel like there needs to be another type of rule, such as a **Style** rule, that's disabled by default, and that can be enforced and configured appropriately for those who wish to have a consistent methodology in their templates. In this case we could have a style rule that offers two modes:
- 'separated-rules' enforces all security group rules to be declared in `AWS::EC2::SecurityGroupIngress` and `AWS::EC2::SecurityGroupEgress` resources, and failing with an **Error** (or **Style Error**) if any rules are declared in the `AWS::EC2::SecurityGroup` resources.
- 'integrated-rules' enforces all security groul rules to be declared in `AWS::EC2::SecurityGroup` resources, and failing with an **Error** (or **Style Error**) if any `AWS::EC2::SecurityGroupIngress` and `AWS::EC2::SecurityGroupEgress` resourxes are declared.

The rule would likely have to be enabled and configured in the config file.

**Comment by kddejong** [comment]
@tro95 thanks for the feedback.  Good call there.  I don't want to enforce much of styling without making it configurable.  What do you think of the **info** level rules?  

For the first one I want to make sure they aren't using both.   From some testing I have been doing its mostly because of updating stacks and not on the initial creation.  Results can vary as you remove or add items and where you add them since both sections are affecting the same rules.  We could additionally make it configurable so someone could force one or the other or both (I suppose).

**Comment by Tro95** [comment]
I like the **Info** rules, however I feel like a separate **Style** rule type makes sense, allowing for the **Info** rules to remain non-blocking and informative, and the **Style** rules to allow blocking failures when the template does not conform. I think some of the **Info** rules could easily be moved over, such as I1022 or I3011 (and personally I'd love I3011 to cause blocking failures).

I'd imagine configuring rules like this:
```yml
templates:
- test/fixtures/templates/good/**/*.yaml
ignore_templates:
- codebuild.yaml
include_checks:
- I
- S3011:
    require_deletion_policy: true
    require_update_policy: false
- S1022:
    error_type: 'warning'
```

Staying on topic, I'd imagine the first rule makes sense as a **Style** rule, and the second as an **Info**. I'm happy to take the style rule discussion to another issue if you'd like.

**Comment by atkinsonm** [comment]
Agree with categorization as a style type of finding. I regularly use this pattern because the property saves lines of code but sometimes the separate resource is needed to avoid circular dependencies. Both are perfectly legal ways to write resources and have zero practical impacts. It comes down to stylistic choice.

Similar situation for the `AWS::IAM::Role` property `Policies` vs `AWS::IAM::Policy` resources.

**Comment by PikaBlue107** [comment]
I'd like to request adding the following rule, too:

- for SecurityGroupEgress: You must specify a destination security group (DestinationPrefixListId or DestinationSecurityGroupId) or a CIDR range (CidrIp or CidrIpv6). (from [AWS CFn Docs on AWS::EC2::SecurityGroupEgress](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-ec2-securitygroupegress.html#cfn-ec2-securitygroupegress-cidrip))
- for SecurityGroupIngress: You must specify a source security group (SourcePrefixListId or SourceSecurityGroupId) or a CIDR range (CidrIp or CidrIpv6). (from [AWS CFn Docs on AWS::EC2::SecurityGroupIngress](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-ec2-securitygroupingress.html#cfn-ec2-securitygroupingress-cidrip))

I'm a bit new to the project; is there a precedent for these "requires any one of these properties" rules?

**Comment by kddejong** [comment]
I think there is a miss on `SecurityGroupIngress` for the resource handler as it should still allow `SourceSecurityGroupName`.  I've created an internal issue for this one. 


I agree.  Working on this.

**Comment by kddejong** [comment]
The `requiredXor` constraints for SecurityGroupIngress (source) and SecurityGroupEgress (destination) have been implemented. The style rules for not mixing inline vs separate SG resources are low priority and debatable per the discussion. Closing this out.
